timemachine/cpp/src/context.cu

```diff
@@ -47,19 +61,60 @@ void Context::add_observable(Observable *obs) {
     this->observables_.push_back(obs);
 }
 
-void Context::multiple_steps(std::vector<double> lambda_schedule) {
-    for(auto lambda : lambda_schedule) {
-        this->_step(lambda);
+std::vector<double> Context::multiple_steps(
+    const std::vector<double> &lambda_schedule,
+    int store_du_dl_freq) {
+    unsigned long long *d_du_dl_buffer = nullptr;
+    // try catch block is to deal with leaks in d_du_dl_buffer
+    if(store_du_dl_freq==0) {
+        store_du_dl_freq = lambda_schedule.size();
     }
-    cudaDeviceSynchronize();
+    int buffer_size = (lambda_schedule.size() + store_du_dl_freq - 1) / store_du_dl_freq;
+    try {
+        // indicator so we can set it to a default arg.
+        gpuErrchk(cudaMalloc(&d_du_dl_buffer, buffer_size*sizeof(*d_du_dl_buffer)));
+        gpuErrchk(cudaMemset(d_du_dl_buffer, 0, buffer_size*sizeof(*d_du_dl_buffer)));
+        for(int i=0; i < lambda_schedule.size(); i++) {
+            // decide if we need to store the du_dl for this step
+            unsigned long long *du_dl_ptr = nullptr;
+            if(i % store_du_dl_freq == 0) {
```

Comment:
Is there any concern that lambda_schedule.size() -1 % != 0? IE Should the last step report du_dl?

Reply:
This is a great question, and often trips people up. But it's actually intended.

Running for 10 steps with a `store_du_dl_freq` of 5 should produce exactly 2 du_dls. We do not trigger the computations for `du/dx, du/dl` etc. unless we are in a step call.

This is actually why @schmolly  was running for 1001 and 10001 steps for parts of her code.

This is because of how the integrators currently work,

```
stp = step
x_t = coords at time t
ddl = if we compute du/dl
stp   0  1  2  3  4  5  6  7  8  9
x_t  0->1->2->3->4->5->6->7->8->9->10
ddl? u              u
```
The `du/dl` for `x_t=10` would be generated on `stp=10`, which is the 11th step in a 0-idx'd system.